jetty-core/jetty-util/src/test/java/org/eclipse/jetty/util/CallbackTest.java

```diff
@@ -0,0 +1,339 @@
+
+    @Test
+    public void testNestedSuccess() throws Exception
+    {
+        TestAbstractCB callback = new TestAbstractCB();
+        Callback nested = new Callback.Nested(callback);
+        nested.succeeded();
+        assertTrue(callback._completed.await(1, TimeUnit.SECONDS));
+        assertThat(callback._completion.getReference(), Matchers.nullValue());
+        assertTrue(callback._completion.isMarked());
+    }
+
+    @Test
+    public void testNestedFailure() throws Exception
+    {
+        Throwable failure = new Throwable();
+        TestAbstractCB callback = new TestAbstractCB();
+        Callback nested = new Callback.Nested(callback);
+        nested.failed(failure);
+        assertTrue(callback._completed.await(1, TimeUnit.SECONDS));
+        assertThat(callback._completion.getReference(), Matchers.sameInstance(failure));
+        assertTrue(callback._completion.isMarked());
+    }
+
+    @Test
+    public void testNestedAbortSuccess() throws Exception
+    {
+        TestAbstractCB callback = new TestAbstractCB();
+        Callback nested = new Callback.Nested(callback);
+
+        Throwable abort = new Throwable();
+        nested.abort(abort);
+        assertFalse(callback._completed.await(100, TimeUnit.MILLISECONDS));
+        assertThat(callback._completion.getReference(), Matchers.sameInstance(abort));
+        assertFalse(callback._completion.isMarked());
+
+        nested.succeeded();
+        assertThat(callback._completion.getReference(), Matchers.sameInstance(abort));
+        assertThat(callback._completed.getCount(), is(0L));
+    }
+
+    @Test
+    public void testNestedAbortFailure() throws Exception
+    {
+        TestAbstractCB callback = new TestAbstractCB();
+        Callback nested = new Callback.Nested(callback);
+
+        Throwable abort = new Throwable();
+        nested.abort(abort);
+        assertFalse(callback._completed.await(100, TimeUnit.MILLISECONDS));
+        assertThat(callback._completion.getReference(), Matchers.sameInstance(abort));
+        assertFalse(callback._completion.isMarked());
+
+        Throwable failure = new Throwable();
+        nested.failed(failure);
+        assertThat(callback._completion.getReference(), Matchers.sameInstance(abort));
+        assertFalse(ExceptionUtil.areNotAssociated(abort, failure));
+        assertThat(callback._completed.getCount(), is(0L));
+    }
+
+    @Test
+    public void testAbortingWrappedByLegacyCallback() throws Exception
+    {
+        TestAbstractCB callback = new TestAbstractCB();
+        Callback legacyCb = new Callback()
+        {
+            @Override
+            public void succeeded()
+            {
+                callback.succeeded();
+            }
+
+            @Override
+            public void failed(Throwable cause)
+            {
+                callback.failed(cause);
+            }
+        };
+
+        Throwable abort = new Throwable();
+        legacyCb.abort(abort);
+
+        // Abort is seen as failure
```

Comment:
I'm very worried about that behavior, if it ever happens to some user they're going to have a very hard time troubleshooting related problems.

Reply:
But if abort throws an `UnsupportedOperationException` then an app might catch it and try to abort again.  i.e. abort is used in fail safe shit-happened code, so I don't think it should ever throw.

I think valid options are:
 + noop
 + noop + warn
 + fail
 + fail + warn